File: marketdna-backend/app/models/randomness.py
from pydantic import BaseModel


class LuckSkillResult(BaseModel):
    luck_score: float
    luck_label: str
    consistency_score: float          # % positive months
    outcome_dispersion: float         # IQR of 1000 random start returns (%)
    pct_positive_starts: float        # % of random starts with positive outcome
    rolling_cv: float                 # avg coefficient of variation across windows (%)


class ConcentrationResult(BaseModel):
    total_return_pct: float
    top_1_contribution: float
    top_5_contribution: float
    top_10_contribution: float
    top_20_contribution: float
    rcr: float                        # Return Concentration Ratio (top-10 / total, %)
    rcr_label: str
    return_actual_pct: float
    return_minus_best_1: float
    return_minus_best_5: float
    return_minus_best_10: float
    return_minus_best_20: float
    concentration_curve: list[float]  # cumulative % contribution, sorted best-day first
    top_days: list[dict]              # [{date, return_pct}] for top 20 days


class FragilityResult(BaseModel):
    fragility_score: float
    fragility_label: str
    worst_period_dependence: float    # max 1y period / total return (%)
    dd_recovery_dependence: float     # % time spent in drawdown
    monte_carlo_p10_cagr: float
    monte_carlo_p50_cagr: float
    monte_carlo_p90_cagr: float
    pct_shuffles_positive: float
    path_iqr: float                   # IQR of shuffled CAGRs (%)
    edge_persistence: float           # % of rolling 252d periods with positive return
    bull_regime_return: float
    bear_regime_return: float
    regime_dependence_score: float
    mc_histogram: list[dict]          # [{cagr_pct, count}] for distribution chart


class RandomnessReport(BaseModel):
    symbol: str
    period_start: str
    period_end: str
    years: float
    cagr: float
    luck_skill: LuckSkillResult
    concentration: ConcentrationResult
    fragility: FragilityResult
    verdict: str
    key_risk: str
    computed_at: str

File: marketdna-web/src/types/randomness.ts
export interface LuckSkillResult {
  luck_score: number
  luck_label: string
  consistency_score: number
  outcome_dispersion: number
  pct_positive_starts: number
  rolling_cv: number
}

export interface ConcentrationResult {
  total_return_pct: number
  top_1_contribution: number
  top_5_contribution: number
  top_10_contribution: number
  top_20_contribution: number
  rcr: number
  rcr_label: string
  return_actual_pct: number
  return_minus_best_1: number
  return_minus_best_5: number
  return_minus_best_10: number
  return_minus_best_20: number
  concentration_curve: number[]
  top_days: { date: string; return_pct: number }[]
}

export interface FragilityResult {
  fragility_score: number
  fragility_label: string
  worst_period_dependence: number
  dd_recovery_dependence: number
  monte_carlo_p10_cagr: number
  monte_carlo_p50_cagr: number
  monte_carlo_p90_cagr: number
  pct_shuffles_positive: number
  path_iqr: number
  edge_persistence: number
  bull_regime_return: number
  bear_regime_return: number
  regime_dependence_score: number
  mc_histogram: { cagr_pct: number; count: number }[]
}

export interface RandomnessReport {
  symbol: string
  period_start: string
  period_end: string
  years: number
  cagr: number
  luck_skill: LuckSkillResult
  concentration: ConcentrationResult
  fragility: FragilityResult
  verdict: string
  key_risk: string
  computed_at: string
}
